Zend/zend_alloc.c

```diff
@@ -318,19 +362,29 @@ struct _zend_mm_huge_list {
 #define ZEND_MM_PAGE_ADDR(chunk, page_num) \
 	((void*)(((zend_mm_page*)(chunk)) + (page_num)))
 
-#define _BIN_DATA_SIZE(num, size, elements, pages, x, y) size,
+#define _BIN_DATA_SIZE(num, size, elements, pages, x, y) \
+	/* Need two words for free slot pointer and shadow */ \
+	MAX(size, sizeof(zend_mm_free_slot*)*2)
+#define _BIN_DATA_SIZE_C(num, size, elements, pages, x, y) \
+	_BIN_DATA_SIZE(num, size, elements, pages, x, y),
 static const uint32_t bin_data_size[] = {
-	ZEND_MM_BINS_INFO(_BIN_DATA_SIZE, x, y)
+	ZEND_MM_BINS_INFO(_BIN_DATA_SIZE_C, x, y)
 };
 
-#define _BIN_DATA_ELEMENTS(num, size, elements, pages, x, y) elements,
+#define _BIN_DATA_ELEMENTS(num, size, elements, pages, x, y) \
+	/* Adjusting size requires adjusting elements */ \
+	(elements / (_BIN_DATA_SIZE(num, size, elements, pages, x, y) / size))
+#define _BIN_DATA_ELEMENTS_C(num, size, elements, pages, x, y) \
+	_BIN_DATA_ELEMENTS(num, size, elements, pages, x, y),
 static const uint32_t bin_elements[] = {
-	ZEND_MM_BINS_INFO(_BIN_DATA_ELEMENTS, x, y)
+	ZEND_MM_BINS_INFO(_BIN_DATA_ELEMENTS_C, x, y)
 };
 
-#define _BIN_DATA_PAGES(num, size, elements, pages, x, y) pages,
+#define _BIN_DATA_PAGES(num, size, elements, pages, x, y) pages
+#define _BIN_DATA_PAGES_C(num, size, elements, pages, x, y) \
+	_BIN_DATA_PAGES(num, size, elements, pages, x, y),
 static const uint32_t bin_pages[] = {
-	ZEND_MM_BINS_INFO(_BIN_DATA_PAGES, x, y)
+	ZEND_MM_BINS_INFO(_BIN_DATA_PAGES_C, x, y)
 };
```

Comment:
It would be great to avoid these changes.

Reply:
I've updated the branch to avoid changing these: https://github.com/php/php-src/pull/14054/commits/5c29ad5367c9023ea21f633ce3cc5f36ba410e32. However, this needs more special cases than updating the `bin_` tables as I did before to reflect the effective size and elements.